File: src/the-pragmatic-programmer-2nd/java/Misc.java
/***
 * Excerpted from "The Pragmatic Programmer, 20th Anniversary Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/tpp20 for more book information.
***/
package com.pragprog.util;

import java.io.*;
import java.util.Enumeration;
import java.util.Vector;
import java.util.StringTokenizer;
import java.awt.*;
import java.awt.event.*;
import com.pragprog.util.Assert;

/**
  * This file contains a number of small Java examples that
  * didn't warrant their own classes.
  */

public class Misc {

  FileInputStream ipstream;

  public void open_passwd() throws FileNotFoundException {

    // This may throw FileNotFoundException...
    ipstream = new FileInputStream("/etc/passwd");
    
    // ...
  }

  public boolean open_user_file(String name) 
    throws FileNotFoundException {
    
    File f = new File(name);
    
    if (!f.exists()) {
      return false;
    }
    
    ipstream = new FileInputStream(f);
    return true;
  }

  public static void loopinv() {
    int arr[] = new int[10];

    int m = arr[0];   // example assumes arr.length > 0
    int i = 1;
    
    // Loop invariant: m = max(arr[0:i-1])
    while (i < arr.length) {
      m = Math.max(m, arr[i]);
      i = i + 1;
    }
  }

  public static void debug(String s) throws IOException {
    FileWriter fw = new FileWriter("debug.log", true);
    fw.write(s);
    fw.flush();
    fw.close();
  }

  public void testListener() {
    Button b = new Button("Sample");
    b.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        System.err.println ("Got button press");
      }
    });
  }

  public void testStringTokenizer() {
    StringTokenizer st1 = new StringTokenizer("this is a test");
    StringTokenizer st2 = new StringTokenizer("this test will work");

    while (st1.hasMoreTokens() && st2.hasMoreTokens()) {
      System.out.println(st1.nextToken());
      System.out.println(st2.nextToken());
    }
  }

  //
  // From Law of Demeter.
  //

  // 
  // Some dummy definitions so the examples will compile...
  //
  public class Money {
    public String printFormat() { return "$0.00"; }
  }

  public class BankAccount {
    public void printBalance() {};
    public Money getBalance() {
      return null;
    }
  }

  public void printToScreen(String s) {};

  public void showBalance(BankAccount acct) {
    Money amt = acct.getBalance();
    printToScreen(amt.printFormat());
  }

  public class Other {
  void showBalance(BankAccount b) {
    b.printBalance();
  }
  }

  //
  // Dummy definitions for the Colada example.
  //
  public class Blender {
    public void addIngredients(Enumeration e) {};
  }

  public class Colada {
    private Blender myBlender;
    private Vector myStuff;

    public Colada() {
      myBlender = new Blender();
      myStuff = new Vector();
    }
    private void doSomething() {
      myBlender.addIngredients(myStuff.elements());
    }
  }

  //
  // A test harness.
  //
  public final static void main(String args[]) {
    try {
      debug("A test string\n");
    } catch (Exception e) {
      System.out.println("Exception caught: " + e.getMessage());
    }
  }
}

File: src/the-pragmatic-programmer-2nd/c/misc.c
/***
 * Excerpted from "The Pragmatic Programmer, 20th Anniversary Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/tpp20 for more book information.
***/
/** 
  * This file contains a number of different c examples, none of which
  * is large enough to warrant it's own file.  
  */

#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>
#include <strings.h>
#include <math.h>

#include "tests.h"


extern char *optarg;
int exitcode = 0;

/**
 * This redefinition of 'exit' is a nasty little hack for testing purposes
 */

#define exit(n) exitcode = n

#define CHECK(LINE, EXPECTED)               \
  { int rc = LINE;                          \
    if (rc != EXPECTED)                     \
        ut_abort(__FILE__, __LINE__, #LINE,  rc, EXPECTED); }


void ut_abort(char *file, int ln, char *line, int rc, int exp) {
  fprintf(stderr, "%s line %d\n'%s': expected %d, got %d\n", 
                  file, ln, line, exp, rc);
  exit(1);
}

#undef exit

/************************************************************************
 * From algorithm_speed.tip
 ***********************************************************************/

typedef struct node Node;
struct node {
  Node *left;
  Node *right;
};

void getNodeAsString(const Node *node, char *buffer) {}

void printTree(const Node *node) {
  char buffer[1000];
  
  if (node) {
    printTree(node->left);
    
    getNodeAsString(node, buffer);
    puts(buffer);

    printTree(node->right);
  }
}

void printTree2(const Node *node) {
  char buffer[1000];
  
  while (node) {
    if (node->left) printTree(node->left);
    
    getNodeAsString(node, buffer);
    puts(buffer);
    
    node = node->right;
  }
}


void printTreePrivate(const Node *node, char *buffer) {
  if (node) {
    printTreePrivate(node->left, buffer);
    
    getNodeAsString(node, buffer);
    puts(buffer);

    printTreePrivate(node->right, buffer);
  }
}

void newPrintTree(const Node *node) {
  char buffer[1000];

  printTreePrivate(node, buffer);
}


/************************************************************************
 * From algorithm_speed.tip
 ***********************************************************************/

void linsearch() {
  int result;
  int n = 0;
  int i;
  int value = 0;
  int array[1];

  array[0] = 0;

  /* look for 'value' in 'array', returning either
   * an index or -1 if not found
   */
  
  result = -1;

  for (i = 0; i < n; i++) {
    if (array[i] == value) {
      result = i;
      break;
    }
  }
}

/* Truncate string to its last maxlen chars */

void string_tail(char *string, int maxlen) {
  int len = strlen(string);
  if (len > maxlen) {
    strcpy(string, string + (len - maxlen));
  }
}

/************************************************************************
 * From concurrency.tip
 ************************************************************************/


void bad_strtok() {
char buf1[BUFSIZ];
char buf2[BUFSIZ];
char *p, *q;

strcpy(buf1, "this is a test");
strcpy(buf2, "this ain't gonna work");

p = strtok(buf1, " ");
q = strtok(buf2, " ");
while (p && q) {
  printf("%s %s\n", p, q);
  p = strtok(NULL, " ");
  q = strtok(NULL, " ");
}
}

 /***********************************************************************/
void usage() {
  fprintf(stderr,"Usage: \n");
  fprintf(stderr,"       string_tail: -c maxlen -s string -t -e expected\n");
  fprintf(stderr,"       CHECK: -C\n");
  exit(1);
}


/**
  * Test harness for this code
  */
int main(int argc, char **argv)
{
  int c;
  int count = 10;  
  char *string = "TEST";
  char *expected = "???";
  struct stat statBuf;

  while ((c = getopt(argc, argv, "tc:s:e:Ca:b:")) != EOF) {
    switch(c) {

      /* String_tail tests */

    case 'c':  /* Get count */
      count = atoi(optarg);
      break;

    case 'e':
      expected = optarg;
      break;

    case 's':  /* Get string */
      string = optarg;
      break;

    case 't':  /* Run Truncate */
      string_tail(string, count);
      if (strcmp(string, expected) != 0) 
      fail("string_tail: expected %s got %s", expected, string);
      exit(0);
      break;

      /* CHECK tests */
      /* nasty hack to stop it exiting */

    case 'C':
      /* this one succeeds */
      CHECK(stat("/tmp", &statBuf), 0);
      if (exitcode != 0) fail("CHECK(/tmp) returned non-zero");

      fprintf(stderr,
    "Expect the message 'stat(\"/tmpxx\", &statBuf)': expected 0, got -1'\n");
      CHECK(stat("/tmpxx", &statBuf), 0);
      if (exitcode != 1) fail("CHECK(/tmpxx) expected 1, got %d", exitcode);
      exit(0);
      break;

    default:
      usage();
      exit(1);
    }
  }
  usage();
  exit(1);
}


